commcare-cloud/commcare_cloud/environment/schemas/postgresql.py

```diff
@@ -45,7 +114,65 @@ class DBOptions(jsonobject.JsonObject):
     password = jsonobject.StringProperty()
     options = jsonobject.DictProperty(unicode)
     django_alias = jsonobject.StringProperty()
-    shards = jsonobject.ListProperty(int, exclude_if_none=True)  # [start, end] pair
     django_migrate = jsonobject.BooleanProperty(default=True)
     query_stats = jsonobject.BooleanProperty(default=False)
     create = jsonobject.BooleanProperty(default=True)
+
+
+class MainDBOptions(DBOptions):
+    name = constants.commcarehq_main_db_name
+    django_alias = 'default'
+
+
+class FormplayerDBOptions(DBOptions):
+    name = constants.formplayer_db_name
+    django_alias = None
+
+
+class UcrDBOptions(DBOptions):
+    name = constants.ucr_db_name
+    django_alias = 'ucr'
+    django_migrate = False
+
+
+class SynclogsDBOptions(DBOptions):
+    name = constants.synclogs_db_name
+    django_alias = 'synclogs'
+
+
+class FormProcessingConfig(jsonobject.JsonObject):
+    _allow_dynamic_properties = False
+    proxy = jsonobject.ObjectProperty(lambda: FormProcessingProxyDBOptions, required=True)
+    partitions = jsonobject.DictProperty(lambda: StrictPartitionDBOptions, required=True)
+
+    @classmethod
+    def wrap(cls, data):
+        for i, (django_alias, db) in enumerate(data['partitions'].items()):
+            db['django_alias'] = db.get('django_alias', django_alias)
+            db['name'] = db.get('name', 'commcarehq_{}'.format(db['django_alias']))
+        self = super(FormProcessingConfig, cls).wrap(data)
+        return self
+
+    def get_db_list(self):
+        return [self.proxy] + sorted(self.partitions.values(),
+                                     key=lambda db: alphanum_key(db.django_alias))
+
+
+class FormProcessingProxyDBOptions(DBOptions):
+    name = constants.form_processing_proxy_db_name
+    django_alias = 'proxy'
+
+
+class PartitionDBOptions(DBOptions):
+    shards = jsonobject.ListProperty(int, exclude_if_none=True)  # [start, end] pair
```

Comment:
not for this PR but could be worth validating the shard config too. We do validate it in HQ (https://github.com/dimagi/commcare-hq/blob/d8fdc0e5b2f7a3200ea743da60be8f808ddd8a60/corehq/sql_db/config.py#L74-L93) but would be nice to know it's valid at this stage already.

Reply:
Yeah good point. When I do it, I'm inclined to copy-paste-modify rather than try to make the two codebases have this in some shared library. I think it's not only just easier, but also the "right" thing to do, in that this code isn't meant to change, so if you _are_ changing something in the code in CommCare HQ, it's not clear that you also want to be changing it in commcare-cloud—for example if we change the format a bit, it may require a two-step rollout where the application code accepts a new format and then the config management tool produces the new format some time later when we're sure everyone's up to date and/or based on some kind of versioning scheme we don't have yet. Thus in my estimation, having the code be shared is as likely to create coupling we then have to work around as it is to save us work maintaining two copies of similar code. Let me know what you think here.

Anyway, I'll do this shortly in a followup PR.

Reply:
Yea, agree with that sentiment. Seems fine for now.